anomalib/deploy/inferencers/base.py

```diff
@@ -0,0 +1,172 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Dict, Optional, Tuple, Union, cast
+
+import cv2
+import numpy as np
+from omegaconf import DictConfig, OmegaConf
+from torch import Tensor
+
+from anomalib.data.utils import read_image
+from anomalib.utils.normalization.cdf import normalize as normalize_cdf
+from anomalib.utils.normalization.cdf import standardize
+from anomalib.utils.normalization.min_max import normalize as normalize_min_max
+from anomalib.utils.post_process import superimpose_anomaly_map
+
+
+class Inferencer(ABC):
+    """Abstract class for the inference.
+
+    This is used by both Torch and OpenVINO inference.
+    """
+
+    @abstractmethod
+    def load_model(self, path: Union[str, Path]):
+        """Load Model."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def pre_process(self, image: np.ndarray) -> Union[np.ndarray, Tensor]:
+        """Pre-process."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def forward(self, image: Union[np.ndarray, Tensor]) -> Union[np.ndarray, Tensor]:
+        """Forward-Pass input to model."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def post_process(self, predictions: Union[np.ndarray, Tensor], meta_data: Optional[Dict]) -> np.ndarray:
+        """Post-Process."""
+        raise NotImplementedError
+
+    def predict(
+        self, image: Union[str, np.ndarray, Path], superimpose: bool = True, meta_data: Optional[dict] = None
+    ) -> np.ndarray:
+        """Perform a prediction for a given input image.
+
+        The main workflow is (i) pre-processing, (ii) forward-pass, (iii) post-process.
+
+        Args:
+            image (Union[str, np.ndarray]): Input image whose output is to be predicted.
+                It could be either a path to image or numpy array itself.
+
+            superimpose (bool): If this is set to True, output predictions
+                will be superimposed onto the original image. If false, `predict`
+                method will return the raw heatmap.
+
+        Returns:
+            np.ndarray: Output predictions to be visualized.
+        """
+        if meta_data is None:
+            if hasattr(self, "meta_data"):
+                meta_data = getattr(self, "meta_data")
+            else:
+                meta_data = {}
+        if isinstance(image, (str, Path)):
+            image = read_image(image)
+        meta_data["image_shape"] = image.shape[:2]
+
+        processed_image = self.pre_process(image)
+        predictions = self.forward(processed_image)
+        anomaly_map, _ = self.post_process(predictions, meta_data=meta_data)
```

Comment:
I'm guessing this would cause some issues for classification-type algorithms. Maybe we can think of a design that supports both segmentation and classification algos.